### PR TITLE
Automatic GitHub action updates

### PR DESCRIPTION
**Pull request by infinisil**
This is an alternative to #39. It runs the [dependabot CLI](https://github.com/dependabot/cli) locally in the existing weekly update CI action (along with some jq magic). <s>Yet untested in CI, but </s> It also works locally, which is really nice!

Best reviewed commit-by-commit, there's some Nix changes in the first 2 commits.

---

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

### CODE REVIEW

**Comment by infinisil** [comment]
I renamed the `check` job to `build` because it's more accurate considering what it does. I'll have to update the branch protection to change the checks which are required before we can merge it.

**Comment by infinisil** [comment]
Nice: https://github.com/NixOS/nixpkgs-check-by-name/actions/runs/8636542602/attempts/1#summary-23676696374

**Review by willbush — APPROVED** [review]

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -14,3 +14,23 @@ jobs:
...
+      - name: test update script
+        run: |
+          nix-build -A autoPrUpdate
+          {
+            result/bin/auto-pr-update .
+            echo ""
+            echo '```diff'
+            git diff
+            echo '```'
+          } > $GITHUB_STEP_SUMMARY
```
Comment (philiptaron) [inline .github/workflows/main.yml:34]: This looks super nice.

File: default.nix [hunk 2]
```diff
@@ -32,8 +33,12 @@ let
 
 
   results = {
-    build = pkgs.callPackage ./package.nix {
-      inherit nixpkgsLibPath initNix runtimeExprPath testNixpkgsPath;
+    # We're using this value as the root result. By default, derivations expose all of their
+    # internal attributes, which is very messy. We prevent this using lib.lazyDerivation
+    build = lib.lazyDerivation {
```
Comment (philiptaron) [inline default.nix:38]: TIL about [`lazyDerivation`](https://noogle.dev/f/lib/derivations/lazyDerivation) 👀

**Comment by infinisil** [comment]
Awesome!

Not sure if this needs to be documented anywhere since nobody will actually need to run this locally, but to make it work locally you need a Docker daemon running, which on NixOS means something like this:

```nix
{
  virtualisation.docker.enable = true;
  networking.firewall.trustedInterfaces = [ "docker0" ];
  users.users.me.extraGroups = [ "docker" ];
}
```

File: scripts/update-github-actions.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+set -euo pipefail
+
+REPO_ROOT=$1
+
+echo -e "<details><summary>GitHub Action updates</summary>\n\n"
+
+# Each dependabot update call tries to update all dependencies,
+# but the resulting files are output individually for each (with the intention of creating a PR for each).
+# We want to have all changes together though, so we just repeatedly take the first one
+# until there's none anymore (-e jq flag)
```
Comment (philiptaron) [inline scripts/update-github-actions.sh:16]: That's not quite what `-e` does by my reading of [the manual](https://jqlang.github.io/jq/manual/#invoking-jq):

> > `--exit-status / -e:`
>
> Sets the exit status of jq to 0 if the last output value was neither false nor null, 1 if the last output value was either false or null, or 4 if no valid result was ever produced. Normally jq exits with 2 if there was any usage problem or system error, 3 if there was a jq program compile error, or 0 if the jq program ran.

**Comment by willbush** [comment]
We're you using https://github.com/nektos/act ? (it depends on docker)

File: scripts/update-github-actions.sh [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+while
+  # Unused argument would be the remote GitHub repo, which is not used if we pass --local
+  create_pull_request=$(LOCAL_GITHUB_ACCESS_TOKEN=$(gh auth token) \
+    dependabot update github_actions this-argument-is-unused --local "$REPO_ROOT" \
+    | jq -ecs 'map(select(.type == "create_pull_request")) | .[0].data')
```
Comment (philiptaron) [inline scripts/update-github-actions.sh:21]: Wow, this is one heck of a pipeline.

I might:

1. Extract `gh auth token` as it won't change in the loop
2. Use the longer form version of `jq` arguments (`--exit-status`, `--compact-output`, `--slurp`)

It's serviceable as is, though.

**Comment by infinisil** [comment]
Nah I'm running it locally by doing what the action does: https://github.com/NixOS/nixpkgs-check-by-name/blob/44d5e2e905037c4614095228d2b9b91eaa119cf2/.github/workflows/update.yml#L19-L20

Or within the current directory:
```
$ nix-build -A autoPrUpdate
$ result/bin/auto-pr-update . > body 
```

File: scripts/update-github-actions.sh [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  echo ""
+
+  jq -er '."pr-body"' <<< "$create_pull_request"
+  echo '</details>'
+
+  jq -c '."updated-dependency-files"[]' <<< "$create_pull_request" \
+    | while read -r fileUpdate; do
+      file=$(jq -er '.name' <<< "$fileUpdate")
+      # -j makes sure to not output a trailing newline
+      jq -ejr '.content' <<< "$fileUpdate" > "$REPO_ROOT/$file"
```
Comment (philiptaron) [inline scripts/update-github-actions.sh:36]: `-j` is `--join-output`

File: scripts/update-github-actions.sh [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+# We want to have all changes together though, so we just repeatedly take the first one
+# until there's none anymore (-e jq flag)
+while
+  # Unused argument would be the remote GitHub repo, which is not used if we pass --local
+  create_pull_request=$(LOCAL_GITHUB_ACCESS_TOKEN=$(gh auth token) \
+    dependabot update github_actions this-argument-is-unused --local "$REPO_ROOT" \
+    | jq -ecs 'map(select(.type == "create_pull_request")) | .[0].data')
+do
+  title=$(jq -er '."pr-title"' <<< "$create_pull_request")
+  echo "<details><summary>$title</summary>"
```
Comment (philiptaron) [inline scripts/update-github-actions.sh:24]: In the other scripts, you do: `echo -e '<details><summary>$title</summary>\n\n'` (with two newlines)

**Review by philiptaron — APPROVED** [review]
Dang Silvan, I'm learning too much. This is definitely effortful in a way that's quite impressive.

File: scripts/update-github-actions.sh (hunk 3 above)
Comment (infinisil) [reply scripts/update-github-actions.sh:16]: Good eye, I was kind of lazy with the wording there, will improve that!

File: scripts/update-github-actions.sh (hunk 6 above)
Comment (infinisil) [reply scripts/update-github-actions.sh:24]: I think it's cleaner here, I'll change the other scripts to match this one

File: scripts/update-github-actions.sh (hunk 3 above)
Comment (infinisil) [reply scripts/update-github-actions.sh]: [(diff)](https://github.com/NixOS/nixpkgs-check-by-name/compare/08e962950f372595027bcbf898a012a97b62d96b..815a4ecd3031e43751baf65c0b2722a6346c6486#diff-a6887f32bd809f67897cc1dd3164202dfe5a6d4933b6f736b3ae864e00a70893R17-R32)

File: scripts/update-github-actions.sh (hunk 4 above)
Comment (infinisil) [reply scripts/update-github-actions.sh]: Done :)

**Review by philiptaron — APPROVED** [review]
# 🦺 👍🏻 👀 
Looks good.

**Comment by philiptaron** [comment]
Hrmph. https://github.com/NixOS/nixpkgs-check-by-name/actions/runs/8637321506/job/23679230914

**Comment by philiptaron** [comment]
Lazy derivation didn't work as expected?

**Comment by infinisil** [comment]
Ohh it did work as expected, this is actually great because it shows that we relied on `.version`, which was kind of internal. I'll make a PR to explicitly expose it

**Comment by infinisil** [comment]
Just opened https://github.com/dependabot/cli/issues/301 to hopefully have a better solution eventually :)

**Comment by infinisil** [comment]
Nice: https://github.com/NixOS/nixpkgs-check-by-name/pull/42

**Pull request merged**